### PR TITLE
Add Slack announcements 

### PR DESCRIPTION
**Pull request by mariana0pachon**
- Added another route in the controller to handle a message to the Slack bot like "guests: Welcome to Person 1 and Person 2 from Amica on May 23 2019 at 9:30 pm in Providence"
- Made a table in the database to store each announcement's message, people, company, date and time, and location 
- Added the graphql queries to get the announcements to display in real time on the side carousel and the panel

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  # An event was published to the API
+  eventPublished: Event!
+
+  # Latest weather data retrieved
+  weatherPublished(latitude: Float!, longitude: Float!): Weather!
+}
+
+scalar UnixDateTime
+
+type Weather {
```
Comment (aaddaamm) [inline schema.graphql]: These types being added seem unrelated to the commit was this file generated wholesale at some point?

File: schema.graphql [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  precipProbability: Float!
+  precipType: String!
+  time: UnixDateTime!
+}
+
+type WeatherMinutelyDetail {
+  data: [WeatherMinutelyData!]!
+  icon: String!
+  summary: String!
+}
```
Comment (aaddaamm) [inline schema.graphql]: could we add a newline to the bottom of this file? Not a deal breaker tho

File: schema.graphql (hunk 1 above)
Comment (mariana0pachon) [reply schema.graphql]: This file is automatically generated, not sure if I should still modify it?

Comment (aaddaamm) [reply schema.graphql]: discussed offline. I think it would be cleaner to generate this file without the changes in this PR and make a new commit. then in this commit only include the changes that would be added to this file. This isn't the end of the world though.

File: app/javascript/components/widgets/guests/carousel.jsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                  const count = prev.primaryLocation.dayAnnouncements.length;
+                  result = assocPath(
+                    ['primaryLocation', 'dayAnnouncements', count],
+                    announcementPublished,
+                    result,
+                  );
+                  result.primaryLocation.dayAnnouncements = sortBy(
+                    prop('publishOn'),
+                    result.primaryLocation.dayAnnouncements,
+                  );
```
Comment (cbarber) [inline app/javascript/components/widgets/guests/carousel.jsx]: Hah! That's a weird quirk of assocPath. This would be a good opportunity to compose ramda functions into something like `insertSortedBy(element, prop)(collection)` and then use `over` / `lensPath` to update inside the object store.

These slides seem like a decent overview of curry/compose in ramdajs:
http://slides.com/andydrew/ramda-js-curry-compose#/

The promise stuff is good to know but really relevant here.

We can go into more detail if you have more questions

File: app/javascript/lib/datetime.test.js [hunk 4]
```diff
@@ -9,3 +9,13 @@ describe('#getMostRecentDay', () => {
     expect(getMostRecentDay('Tuesday', today)).toEqual('2019-05-07');
   });
 });
+
+describe('#isDateToday', () => {
```
Comment (cbarber) [inline app/javascript/lib/datetime.test.js]: Nice job adding test coverage!

File: app/javascript/components/widgets/guests/panel.jsx [hunk 5]
```diff
@@ -1,5 +1,92 @@
...
-export default Panel;
+const getPrimaryLocationAnnouncements = gql`
+  {
+    primaryLocation {
+      dayAnnouncements {
+        message
+        people
+        company
+        publishOn
+        id
```
Comment (cbarber) [inline app/javascript/components/widgets/guests/panel.jsx]: Looks like id isn't used here

File: app/javascript/components/widgets/guests/panel.jsx [hunk 6]
```diff
@@ -1,5 +1,92 @@
...
+  <AnnouncementWrapper>
+    <AnnouncementMessage>{announcement.message}</AnnouncementMessage>
+    <AnnouncementTitle>
+      {announcement.people} from {announcement.company}
+    </AnnouncementTitle>
+    <AnnouncementDate>{parseDate(announcement.publishOn)}</AnnouncementDate>
+  </AnnouncementWrapper>
+);
+Announcement.propTypes = {
+  announcement: PropTypes.objectOf.isRequired,
```
Comment (cbarber) [inline app/javascript/components/widgets/guests/panel.jsx]: `shape` for the explicit props here as well

File: app/javascript/components/widgets/guests/carousel.jsx [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+              document: subscribeAnnouncementPublished,
+              updateQuery: (prev, { subscriptionData }) => {
+                if (!subscriptionData.data) {
+                  return prev;
+                }
+                const { announcementPublished } = subscriptionData.data;
+
+                let result = prev;
+
+                if (isToday(announcementPublished.publishOn)) {
```
[the remark this answers is not in the document]
Comment (mariana0pachon) [reply app/javascript/components/widgets/guests/carousel.jsx]: Solved by adding a `pollInterval` of 1 hour to the Query

File: app/javascript/components/widgets/guests/carousel.jsx (hunk 3 above)
Comment (mariana0pachon) [reply app/javascript/components/widgets/guests/carousel.jsx]: Decided to sort when mapping through the array of announcements and generating the components. The subscription now only needs to append the `announcementPublished` to `prev`.

File: app/graphql/types/announcement_type.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+class Types::AnnouncementType < Types::BaseObject
```
[the remark this answers is not in the document]
Comment (mariana0pachon) [reply app/graphql/types/announcement_type.rb]: Moved to the gql query commit instead

File: app/models/announcement.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  validates :publish_on, presence: true
+  validates :people, presence: true
+  validates :event_id, presence: true
+
+  belongs_to :location, inverse_of: :announcements
+
+  scope :with_location, ->(location) { where(location: location) }
+  scope :happen_today, ->(time_zone) do
+    today = Time.now.in_time_zone(time_zone).to_datetime
+    tomorrow = today + 24.hours
```
[the remark this answers is not in the document]
Comment (mariana0pachon) [reply app/models/announcement.rb]: Used ranges today..tomorrow instead, and got rid of to_datetime and beginning_of_day

**Comment by thedelchop** [comment]
@mariana0pachon I know that @cbarber left you some comments but is on PTO this week, SO, if you want to make corrections and ping me I can re-review.

File: app/javascript/components/widgets/guests/panel.jsx [hunk 10]
```diff
@@ -0,0 +1,97 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Query } from 'react-apollo';
+import gql from 'graphql-tag';
+import styled from 'styled-components';
+import { parseDate } from '../../../lib/datetime';
+import { colors, weights, fontSizes, spacing } from '../../../lib/theme';
+
+const LoadingMessage = () => <p>Loading...</p>;
```
Comment (geraldalewis) [inline app/javascript/components/widgets/guests/panel.jsx]: @theroyalblah @mariana0pachon Just pointing out that Simon's work in https://github.com/mojotech/helios2/pull/90 introduces new global `LoadingMessage` and `ErrorMessage` components. If this PR is merged first, Simon can update his PR to update these components. If Simon's PR is merged first, Mariana can update this commit to make use of the new `LoadingMessage` and `ErrorMessage` components. Sound good?

File: app/javascript/components/widgets/guests/carousel.jsx [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+import PropTypes from 'prop-types';
+import { Query } from 'react-apollo';
+import gql from 'graphql-tag';
+import { sortBy, prop, append, lensPath, set } from 'ramda';
+import styled from 'styled-components';
+import icon from '../../../../assets/images/icons/icon-calendar.svg';
+import { parseTime, isToday } from '../../../lib/datetime';
+
+const LoadingMessage = () => <p>Loading...</p>;
+const ErrorMessage = ({ message }) => <p>Error: {message}</p>;
```
Comment (djkirby) [inline app/javascript/components/widgets/guests/carousel.jsx]: Should be able to utilize the changes from #90 at this point

File: app/javascript/components/widgets/guests/panel.jsx [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+`;
+
+const AnnouncementDate = styled.div`
+  width: 50vw;
+  opacity: 0.5;
+  font-size: ${fontSizes.medium};
+  color: ${colors.white};
+`;
+
+const Announcement = ({ announcement }) => (
```
Comment (djkirby) [inline app/javascript/components/widgets/guests/panel.jsx]: Could reduce some repetitiveness by destructuring here,
```tsx
const Announcement = ({ announcement: { message, people, company, publishOn } }) => (
```

File: app/javascript/components/widgets/guests/panel.jsx [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    company: PropTypes.string.isRequired,
+    publishOn: PropTypes.string.isRequired,
+    id: PropTypes.string.isRequired,
+  }).isRequired,
+};
+
+const Guests = () => (
+  <Query query={getPrimaryLocationAnnouncements}>
+    {({ loading, error, data }) => {
+      if (loading) {
```
Comment (djkirby) [inline app/javascript/components/widgets/guests/panel.jsx]: can probably remove this condition since we fall back to it further below

File: app/controllers/web_hooks/slack_controller.rb [hunk 14]
```diff
@@ -23,9 +10,50 @@ def build_create_response
...
+    type, content = event_callback[:event][:text].split(/: /)
+
+    case type
+    when 'guests'
+      content = content.split(/ to | from | on | in /).map(&:strip)
+      fields = %i[message people company publish_on location_id]
+      announcement = Hash[fields.zip(content)]
+      announcement[:announcement_id] = event_callback[:event_id]
+      announcement[:location] = Location.find_by(city_name: announcement[:location_id])
+      Time.zone = announcement[:location][:time_zone]
```
Comment (cbarber) [inline app/controllers/web_hooks/slack_controller.rb]: I think `use_zone` with a block would be a little cleaner and would restore the original Time zone value.

File: app/controllers/web_hooks/slack_controller.rb [hunk 15]
```diff
@@ -23,9 +10,50 @@ def build_create_response
...
+      announcement[:publish_on] = Time.zone.parse(announcement[:publish_on])
+      return announcement
```
Comment (cbarber) [inline app/controllers/web_hooks/slack_controller.rb]: Just `announcement` without a return preferred

File: app/controllers/web_hooks/slack_controller.rb [hunk 16]
```diff
@@ -23,9 +10,50 @@ def build_create_response
...
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  private def parse_announcement(event_callback)
+    # Ex: "guests: Welcome to Roy and Amanda from Under Armour on May 23rd at 11:00 am in Providence"
+    type, content = event_callback[:event][:text].split(/: /)
+
+    case type
+    when 'guests'
+      content = content.split(/ to | from | on | in /).map(&:strip)
```
Comment (cbarber) [inline app/controllers/web_hooks/slack_controller.rb]: Would extracting this as a `handle_guests_command` method resolve the metrics issue above?

File: app/controllers/web_hooks/slack_controller.rb (hunk 14 above)
Comment (mariana0pachon) [reply app/controllers/web_hooks/slack_controller.rb]: Something like this? 
`
    Time.use_zone(announcement[:location][:time_zone]) {
      announcement[:publish_on] = Time.zone.parse(announcement[:publish_on])
    }
`

Comment (cbarber) [reply app/controllers/web_hooks/slack_controller.rb]: Yes that works. rubocop will likely switch it to do/end with a new line

File: app/javascript/components/widgets/guests/carousel.jsx [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    message: PropTypes.string.isRequired,
+    people: PropTypes.string.isRequired,
+    company: PropTypes.string.isRequired,
+    publishOn: PropTypes.string.isRequired,
+    id: PropTypes.string.isRequired,
+  }).isRequired,
+  subscribeToPublishedAnnouncements: PropTypes.func.isRequired,
+};
+
+const Guests = () => (
```
Comment (aaddaamm) [inline app/javascript/components/widgets/guests/carousel.jsx]: can we update this component name to match the file name? also, i think it should be named something like `GuestsCarouselItem` until we come up with a better name.

File: app/javascript/components/widgets/guests/index.jsx [hunk 18]
```diff
@@ -0,0 +1,7 @@
+import Carousel from './carousel';
+import Panel from './panel';
+
+export default {
+  Carousel,
```
Comment (aaddaamm) [inline app/javascript/components/widgets/guests/index.jsx]: when we are exporting we can do it like 
```
export default {
  CarouselItem: GuestsCarouselItem,
  Panel: GuestsPanel,
}
```

File: app/javascript/components/widgets/guests/carousel.jsx (hunk 17 above)
Comment (cbarber) [reply app/javascript/components/widgets/guests/carousel.jsx]: I created a ticket to switch the names from carousel to tab. Might be better to remain consistent for now and switch every thing in one go.

Comment (aaddaamm) [reply app/javascript/components/widgets/guests/carousel.jsx]: oh ok then.

File: app/javascript/components/widgets/guests/index.jsx (hunk 18 above)
Comment (cbarber) [reply app/javascript/components/widgets/guests/index.jsx]: I don't think this intermediary naming has an impact on the dev tools. Was there another reason to prefer `import GuestsCarouselItem from './carousel'` over `import Carousel from './carousel'`

Comment (aaddaamm) [reply app/javascript/components/widgets/guests/index.jsx]: i like the imports to match the names they are exporting. I think it reduces the overhead when trying to understand files like this.
